src/main/java/org/reactome/addlinks/dataretrieval/ensembl/EnsemblBiomartRetriever.java

```diff
@@ -0,0 +1,232 @@
+                if (line.contains("ERROR") ) {
+                    retryCount++;
+                    if (retryCount < MAX_QUERY_RETRIES) {
+                        return retryQuery(queryString, retryCount);
+                    }
+                    // The data does not exist in BioMart for a few species. Frog (X. tropicalis) has UniProt data but not Microarray data;
+                    // Yeast (S. cerevisiae) has Uniprot-SwissProt data but not UniProt-TrEMBL data; Yeast (S. pombe), P. falciparum
+                    // and D. discoideum don't have UniProt or Microarray data, all at time of writing (January 2020).
+//                    throw new Exception("Biomart query failed with message: " + line + "\nThis can happen without issue for certain species (X. tropicalis, S. pombe, S. cerevisiae, P. falciparum) because the data doesn't exist in BioMart");
+                    throw new BioMartQueryException(line +
+                            "\nThis can happen without issue for certain species (X. tropicalis, S. pombe, S. cerevisiae, P. falciparum) " +
+                            "because the data doesn't exist in BioMart");
+                } else {
+                    biomartResponseLines.add(line);
+                }
+            }
+            return biomartResponseLines;
+
+        } else {
+            retryCount++;
+            if (retryCount < MAX_QUERY_RETRIES) {
+                return retryQuery(queryString, retryCount);
+            }
+            throw new HttpException(
+                    "Unable to connect to Biomart (" +
+                            biomartConnection.getResponseCode() + ": " + biomartConnection.getResponseMessage() +
+                            ") with URL: " + queryString
+            );
+        }
+    }
+
+    // Recursive method that retries the BioMart query.
+    private Set<String> retryQuery(String queryString, int retryCount) throws InterruptedException, IOException, HttpException, BioMartQueryException {
+        final long QUERY_SLEEP_DURATION = Duration.ofSeconds(5).toMillis();
+
+        Thread.sleep(QUERY_SLEEP_DURATION);
+        logger.warn("Biomart query failed. Trying again...");
+        return queryBiomart(queryString, retryCount);
+    }
+
+    // Gets the filepath to biomart XML file, used to build biomart query.
+    private String getBiomartXMLFilePath() throws IOException {
+        // Get species JSON config file location, import as JSON object.
+        return this.properties.getProperty("pathToBiomartXML");
+    }
+
+    // Modifies default secondary URL with species name and the query type (for example microarray probe type, or uniprotsptrembl).
+    private String getBiomartIdentifierQuery(String pathToQueryFile, String biomartSpeciesName, String queryId) throws IOException {
+        String biomartQuery = "http://www.ensembl.org/biomart/martservice?query=" + URLEncoder.encode(joinQueryFileLines(pathToQueryFile),"UTF-8");
```

Comment:
For the second argument to `URLEncoder.encode`, you could use `StandardCharsets.UTF_8.toString()`

Reply:
That seems to add a lot of text just to escape a string value?

Reply:
Yes, but if `URLEncoder.encode` ever changes, using a defined constant for the correct encoding will avoid problems in the future if the string literal "UTF-8" becomes invalid. Not that I think this will ever actually happen... but it's a good habit to get into, in general. ;)

Reply:
That's fair - one thing you could do to reduce the text size is a static import of the UTF_8 charset: `import static java.nio.charset.StandardCharsets.UTF_8;`.

Then it would just be `UTF_8.toString()`

It's not that big a deal to leave it as a literal String, though.